### PR TITLE
hierarchy of errors

### PR DESCRIPTION
**Pull request by YeungOnion**
looking for some feedback on possible error hierarchy.
I don't think I did the source error pattern nicely, but I think there's something here worth looking at.

### CODE REVIEW

File: src/distribution/gamma.rs [hunk 1]
```diff
@@ -1,9 +1,78 @@
...
+            DegenerateSuccessCount => Self::DegenerateRate(0.0),
+        }
+    }
+}
+
+/// holds a valid parametrization of the gamma distribution in shape and rate.
+pub struct Parameters {
+    shape: f64,
+    rate: f64,
+}
```
Comment (FreezyLemon) [inline src/distribution/gamma.rs:44]: Can you explain why you added this abstraction? You could also check the parameters inside `Gamma::new` and return a suitable error from there, right?

**Review by FreezyLemon — COMMENTED** [review]
This looks pretty nice. The error contains a lot of information (maybe too much? I can't estimate if this level of detail is needed) and the source annotation from `thiserror` should create better dev feedback for little effort.

There's one thing I don't really understand but I like this in general

File: src/distribution/gamma.rs (hunk 1 above)
Comment (YeungOnion) [reply src/distribution/gamma.rs:44]: Mostly just playing around with having input checking in another type to support the use of infallible `new` methods. 

But more interestingly perhaps it _could_ be a way to get alternative parametrizations #198 as well as `new_from_moments` to do something like,

```rs
let m = Gamma::new(shape, rate).into_moments();
let p = Poisson::new_from_moments(m);
```

But I've lots of thoughts floating around about that none motivated by something I've actually needed thus far. As a solution to #198 could be adding just one more method.

**Comment by YeungOnion** [comment]
#265 will do this without unused complexity. Thanks @FreezyLemon

**Pull request closed without merging**